### PR TITLE
Extend RSA capability

### LINKED ISSUE #16181: Improve RSA Documentation
`encipher_rsa` and `decipher_rsa` docs only informs about encrypting with public key and decrypting with private key. But it is also possible to encrypt message using private key and decrypting it with public key.

The examples may also contain a process of 'signing' a message with private key,
and verifying the message by comparing the original message, with deciphered message with public key.

### PR DESCRIPTION
**Pull request by sylee957**
#### References to other Issues or PRs

Closes #16204
Fixes #16181

#### Brief description of what is fixed or changed

There had been a question of using non-primes or non-distinct primes for `p`, `q` in #16162 

This can be more clearly answered after generalizing RSA to multiprime and multipower cases.

I found that RSA can be generalized for any arbitrary 
where image are distinct primes and image are positive integers.
Which can simply be called prime factors of `n`, and the definition can be generalized for every possible pairs `p, q` for creating the RSA modulus `n`, if the RSA is defined in this point.

And there are two types of RSA that can be roughly categorized into. 
 - First one is a purely multiprime case image where the exponents are 1.
 - Second one is a multipower case  where any of the exponents can be more than 1.

For both cases, the RSA formula can be applied without any much changes, compared to the popular two-distinct prime case. 
 - How to encrypt and decrypt remains same as image and image where `M`, `C` are plaintext and ciphertext, `e`, `d` are the exponents of public and private key, and `n` is the modulus shared by public and private key as described above.
 - And also the way to compute `d` from `e` remains same as image or image.

However, the only difference between multiprime RSA and multipower RSA is that:
 - Multiprime RSA have bijective mapping in the complete residue system image, which is a range of every integers from `0` to `n-1`
 - Multipower RSA may not have bijective mapping in the complete residue system, but in the reduced residue system image, which means integers from `0` to `n-1` and also relatively prime to `n`.

So, if the concept of RSA can be generalized like this, we can consider
 - RSA with distinct primes `p`, `q` : A special case of multiprime RSA
 - RSA with same primes `p`, `p` : A special case of multipower RSA
 - RSA with non-prime `p`, `q` : Compute `n = p*q` and prime factor `n`, then it can be canonicalized either into an another multiprime RSA or a multipower RSA.
 - RSA with a single prime `p` : A trivial case of multiprime RSA, though it may not have a practical means.
 - RSA with a single non-prime `p` : Prime factor `n = p` and consider it as a multiprime or a multipower case.
 - RSA with any arbitrary `p, q, r, ...`

#### Other comments

I've added an option `totient` to control either to use Euler's totient or a Carmichael's totient, because they can yield different `d`. 
And I find that almost every articles I had read are using the exploit `(p-1)*(q-1)` or  `lcm(p-1, q-1)`, which can be fall into the notation of either one of those.

I've also added an option to use chinese remainder theorem to compute the encryption and decryption. It is not really faster in small cases, but it can still be useful to check if such algorithm works in nontrivial way.

<details><summary>A simple benchmark</summary>
<p>

```python
from sympy import *
from sympy.crypto import *

p = int(
    '1017715760715424506802386150369308212090648714372506228340650154082258' \
    '2262040469998382971671408213646381806971948795002455576544518696289334' \
    '6463841419427008800341257468600224049986260471922572481630144688417254' \
    '7691863941572670973607781363296129091102564212329778330286774412060493' \
    '09220354796014376698325101693')

q = int(
    '2875234235309513287229018152660727588618279324166080507785080175689512' \
    '7977542869729522735531281818618305768362897386687452503402819969112887' \
    '0676414118458442900035778874482624765513861643279666963168221883983361' \
    '9900230658870390289410047618682384959510323941052727960544214828581614' \
    '9368667083114802852804976893')

r = int(
    '1769822925986882577687950073635018683885096193595631013437826189771862' \
    '1867174630675413696948162452252919211380388001711255960731544952198115' \
    '7084370187887650624061033066022458512942411841187478937899723152771600' \
    '8508616411987953604187533538484482056602874796176717264080533196198920' \
    '52000850883994343378882717849')

s = int(
    '6892542843858543102926918223350261102709175506464374238351562364321310' \
    '5828968933955293670749428083531871387944227457184196452829123186515721' \
    '2604266903677599180789896916456120289112752835985022658896697303316882' \
    '0682522007471397760741517873801583103036429058536915050281974382734355' \
    '2098197095520550865360159439'
)

t = int(
    '6903548343345363282055131189236890877977814456871145530154109431487047' \
    '6423226953576968609257479231896350331830698238209105217117290910679774' \
    '8883261493224162414050106920442445896819806600154484448261080082179721' \
    '2913062557142190489325280472987735335273942048057484285020218146265625' \
    '1626522910618936534699566291'
)

e = 65537

puk = rsa_public_key(p, q, r, s, t, e)
prk = rsa_private_key(p, q, r, s, t, e)
```
```python
import cProfile

cProfile.run('decipher_rsa(1000, prk)')
```
```
         6 function calls in 0.255 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.255    0.255 <string>:1(<module>)
        1    0.000    0.000    0.255    0.255 crypto.py:1732(_encipher_decipher_rsa)
        1    0.000    0.000    0.255    0.255 crypto.py:1804(decipher_rsa)
        1    0.000    0.000    0.255    0.255 {built-in method builtins.exec}
        1    0.254    0.254    0.254    0.254 {built-in method builtins.pow}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```
```python
cProfile.run('decipher_rsa(1000, prk, [p, q, r, s, t])')
```
```
         413 function calls (412 primitive calls) in 0.085 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       38    0.000    0.000    0.000    0.000 <frozen importlib._bootstrap>:1009(_handle_fromlist)
        1    0.000    0.000    0.000    0.000 <frozen importlib._bootstrap>:103(release)
        1    0.000    0.000    0.000    0.000 <frozen importlib._bootstrap>:143(__init__)
        1    0.000    0.000    0.000    0.000 <frozen importlib._bootstrap>:147(__enter__)
        1    0.000    0.000    0.000    0.000 <frozen importlib._bootstrap>:151(__exit__)
```
</p>
</details>

I'll also post the sphinx documentation build here as below. because it is too large, but it may contain some additional information not being noted here

<details><summary>RSA Public Key Doc</summary>
<p>

</p>
</details>

<details><summary>RSA Private Key Doc</summary>
<p>

</p>
</details>

<details><summary>Encipher RSA Doc</summary>
<p>

</p>
</details>

<details><summary>Decipher RSA Doc</summary>
<p>

</p>
</details>

#### Release Notes

- crypto
  - Extended `rsa_public_key` and `rsa_private_key` to handle any arbitrary multiprime RSA or multipower RSA case.
  - Added an optional argument `factors` for `encipher_rsa` and `decipher_rsa`, which allows to use chinese remainder theorem to compute the modulo.
  - Added an optional argument `totient` for `rsa_public_key` and `rsa_private_key`, which allows to control the notation of totient to be used.

### CODE REVIEW

File: sympy/crypto/crypto.py [hunk 1]
```diff
@@ -1383,24 +1396,227 @@ def bifid6_square(key=None):
...
+
+        It can also be possible to form a single-prime RSA by specifying
+        the argument as `p, e`, which can be considered a trivial case
+        of a multiprime RSA.
+
+        Furthermore, it can be possible to form a multi-power RSA by
+        specifying two or more pairs of the primes to be same.
+        However, unlike the two-distinct prime RSA or multi-prime
+        RSA, not every numbers in the complete residue system
+        (`\mathbb{Z}_n`) will be decryptable since the mapping will not
```
Comment (jksuom) [inline sympy/crypto/crypto.py]: What is the mapping that is referred to here?

Comment (sylee957) [reply sympy/crypto/crypto.py]: It's referring to complete residue system to itself for multiprime RSA cases, and reduced residue system to itself for multipower RSA cases

**Comment by jksuom** [comment]
Perhaps it should be explained in the documentation why the original RSA was designed for a modulus that is the product of two big primes. (It is possible to break the code by factoring the modulus. Numbers  with two prime factors are the hardest to factor.)

**Comment by sylee957** [comment]
Further generalizing the non-uniqueness of RSA private key according to each totient notation, I have added a new keyword `index` to give any possible public or private exponents which can be used in the same manner.

Probably, there can be some more things to organize the docs, because parameters section contains too much mathematical details now. However, I wonder if there can be any muh cleaner way to explain such unconventional concepts.

**Comment by oscarbenjamin** [comment]
Looks good

**Pull request merged**